Feature/issue#48 알림 기능 비동기로 변경

Jeonwoojae opened the pull request:
### ✍️ Description
- 리소스를 효율적으로 사용하고 빠른 반응 속도를 위해 이벤트 처리를 비동기로 변경했습니다. 

<br>

### 🎯 Key Changes
- 비동기 작업을 수행하는 메서드에 `@Async` 사용
- 비동기 작업을 수행하기 위한 `Executor` 선언
- 테스트 코드에서 1초 더 기다리도록 변경
  - 해당 방법은 비동기를 동기로 변경하면서 기다리지 않게 되었습니다

<br>

### 💬 To Reviewers
- 각 이벤트를 처리하는 코드가 보기 불편하지 않으신가요?
  - 비동기 작업을 확인하기 위해 로깅 작업을 AOP로 분리하면 어떨까 합니다 
- 테스트 시 기존 스레드가 먼저 끝나면 테스트를 통과할 수 없어 1초를 기다리도록 하고 있는데 좋은 테스트는 아니라고 생각합니다.😢

close #48

Jeonwoojae committed:
feat: 음악 추가 시 알림 생성, 전송 기능 비동기로 변경

- 기본 Executor보다 커스텀 할 수 있는 Executor 사용
- @Async 사용하여 비동기 적용
- Test 코드 수행 시 종료 타이밍이 보장되지 않아 Executor로 종료 대기

minju412 reviewed (approved):
고생하셨습니다!! 👍👍 지난 회의 때 말씀드린 것처럼 로깅 작업은 aop와 filter/interceptor의 선택 기준을 정하시면 좋을 것 같아요! 

그리고 비동기 작업에 대해서 잘은 모르지만 알아보니 비슷한 상황에서 CountDownLatch를 이용한 예제가 있는 것 같아요! 제가 테스트 상황을 잘 이해하지 못한 것일 수 있으니 참고만 부탁드립니다..ㅎㅎ
- [CountDownLatch를 이용해 thread 대기하기](https://1minute-before6pm.tistory.com/30)
- [병행성(Concurrency)을 위한 CountDownLatch](https://imasoftwareengineer.tistory.com/100)

Jeonwoojae committed:
feat: 비동기 작업 로깅 기능 추가

- Executor에 decorate를 override하여 해당 Executor를 작업하는 모든 비동기 작업 스레드 정보, 소요 시간 로깅

Jeonwoojae committed:
test: 비동기 테스트 동기 테스트로 변경

- TestConfiguration을 사용하여 해당 테스트에서 동기 Executor를 사용해 비동기 로직을 동기 로직으로 테스트 할 수 있음

Jeonwoojae committed:
test: Notification 통합 테스트 문제 해결

- 기존 전체 테스트를 한번에 진행하면 데이터 롤백을 하지 못해 테스트를 통과하지 못하는 문제를 매 테스트 마다 springContext를 초기화 하여 해결

Jeonwoojae committed:
refactor: Event 생성 시 ApplicationEvent 상속 삭제

- Spring 4.2부터 ApplicationEvent를 상속 받지 않아도 publish 가능

Jeonwoojae committed:
refactor: Notification 생성 static 팩토리 메서드 사용

- 객체 생성 부분을 캡슐화 할 수 있음

Jeonwoojae committed:
refactor: executor 설정파일 분리

- Application 시작점에 Sl4j를 직접 설정하는 것은 의미상 옳지 않아 표시할 Configuration 분리
- 기존 taskExecutor 네이밍은 예약어가 있어 asyncTaskExecutor로 변경

Nine-JH reviewed (approved):
👍👍

Jeonwoojae commented on src/main/java/com/petitapetit/miml/global/config/AsyncConfig.java:
```diff
@@ -0,0 +1,40 @@
+    @Bean
+    public Executor asyncTaskExecutor() {
+        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
+        executor.setCorePoolSize(4);
```
각 설정에 대한 설명
- CorePoolSize : 최소한으로 유지되는 스레드의 개수
- MaxPoolSize : 스레드 풀에서 생성할 수 있는 최대 스레드의 수
- QueueCapacity : 작업이 스레드 풀의 최대 크기를 초과할 경우 대기열에 저장할 수 있는 작업의 개수

Jeonwoojae replied:
### 동작하는 순서
1. 기본적으로 CorePoolSize만큼 스레드가 일을 처리합니다.
2. 기본 스레드가 처리할 수 있는 작업량을 넘어서면 LinkedBlockingQueue를 생성하여 대기시킵니다.
3. queue도 꽉 차면 MaxPoolSize까지 스레드를 추가로 생성하여 작업을 처리합니다.
